File: csharp/878.nth-magical-number.cs
/*
 * @lc app=leetcode id=878 lang=csharp
 *
 * [878] Nth Magical Number
 */

// @lc code=start
public class Solution {

    private static int GreatestCommonDivisor(int a, int b) {
        if (b == 0) {
            return a;
        }
        return GreatestCommonDivisor(b, a % b);
    }

    public int NthMagicalNumber(int N, int A, int B) {
        int l = A * B / GreatestCommonDivisor(A, B), count = l / A + l / B - 1, i = N % count, left = 0, right = l - 1;
        while (left < right) {
            int mid = left + (right - left) / 2;
            if (mid / A + mid / B < i) {
                left = mid + 1;
            } else {
                right = mid;
            }
        }
        return (int)((((long)l * (long)(N / count)) % 1000000007 + (long)left) % 1000000007);
    }
}
// @lc code=end

File: java/878.nth-magical-number.java
/*
 * @lc app=leetcode id=878 lang=java
 *
 * [878] Nth Magical Number
 */

// @lc code=start
class Solution {

    private static int getGcd(int x, int y) {
        return x == 0 ? y : getGcd(y % x, x);
    }

    public int nthMagicalNumber(int n, int a, int b) {
        long lcm = (long)a * (long)b / (long)getGcd(a, b), l = (long)Math.min(a, b), r = (long)n * (long)Math.max(a, b);
        while (l < r) {
            long m = (l + r) / 2;
            if ((long)n > m / a + m / b - m / lcm) {
                l = m + 1;
            } else {
                r = m;
            }
        }
        return (int)(l % 1000000007);
    }
}
// @lc code=end

